contents/why_julia.md

```diff
@@ -69,32 +74,271 @@ Let's proceed then!
+You would need to change your approach with a substancial loss of simplicity (you would probably need to create different functions with different names).
+
+**Now let's do this in Julia**.
+
+In Julia, we don't have classes but we have structures (`struct`) that are meant to be "structured data": they define the kind of information that is embedded in the structure, that is a set of fields (i.e. "properties" or "attributes" in other languages), and then individual instances (or "objects") can be produced each with its own specific values for the fields defined by the structure.
+They differ from the primitive types that are by default defined already inside the core of Julia.
+Because of this, `structs`s are composite types, and usually defined by the user.
+
+First we'll create an `abstract type` named `Researcher`:
+
+```jl
+sc(
+"""
+abstract type Researcher end
+"""
+)
+```
+
+We proceed, similar as before, by creating two derived `struct`s from the `Researcher` abstract type:
+
+```jl
+sc(
+"""
+struct Linguist <: Researcher
+    name::AbstractString
+    age::Int64
+end
+"""
+)
+```
+
+```jl
+sc(
+"""
+struct ComputerScientist <: Researcher
+    name::AbstractString
+    age::Int64
+end
+"""
+)
+```
+
+The final step is to define two new functions that will behave differently depending on which derived `struct` of `Researcher` are the first or second argument:
+
+```jl
+sco(
+"""
+approaches(li::Linguist, cs::ComputerScientist) = "Hey \$(cs.name), wanna do a paper? We need to use APA style."
+
+approaches(cs::ComputerScientist, li::Linguist) = "Hey \$(li.name), wanna do a paper? We need to use IEE style."
+"""
+)
+```
+
+Finally, let's instantiate our two researchers, `noam` and `judea`:
+
+```jl
+sc(
+"""
+noam = Linguist("Noam Chomsky", 92)
+judea = ComputerScientist("Judea Pearl", 84)
+"""
+)
+```
+
+Again, let's see what Noam Chomsky will say when he approaches Judea Pearl with a paper idea:
+
+```jl
+sco(
+"""
+approaches(noam, judea)
+"""
+)
+```
+
+Perfect! It behaves just as we wanted!
+This is **multiple dispatch** and it is an important feature in Julia.
+Multiple dispatch acts on all arguments of a function and defines function behavior based on all argument's types.
+
+Multiple dispatch is a powerful feature that allows us also to extend existing functions or to define custom and complex behavior to new types.
+To show how this works, we'll use another example.
+Suppose you want to define two new `struct`s based on a animals:
+
+```jl
+sc(
+"""
+struct fox end
+struct chicken end
+"""
+)
+```
+
+I want to define addition for both `fox` and `chicken` new types.
+We proceed by defining a new function signature of the `+` from the `Base` module of Julia^[this is just an example for teaching purposes. Doing something similar as this example will probably result in a [method invalidation](https://julialang.org/blog/2020/08/invalidations/).]:
+
+```jl
+sco(
+"""
+import Base: +
++(F::fox, C::chicken) = "trouble"
```

Comment:
I love this example. It's really funny to read the computer giving the dry output of "trouble" :rofl: 

Maybe, also define `+(C::chicken, C::chicken) = "ok"` or "happy" or so?

Reply:
Yes! defined also in d2c7843